### PR TITLE
allow connection to a peer with unknown PeerId

### PR DESCRIPTION
**Pull request by arnetheduck**
Useful when connecting to a noise peer based on IP address alone

### CODE REVIEW

**Comment by Menduist** [comment]
The new API for this:
```nim
check: (await switch2.connect(switch1.peerInfo.addrs)) == switch1.peerInfo.peerId
```

Since it returns the `PeerId`, it should be pretty obvious what it does, and can't be used unintentionally insteado f the original `connect` (which returns void)

File: libp2p/dial.nim [hunk 1]
```diff
@@ -31,6 +31,13 @@ method connect*(
 
   doAssert(false, "Not implemented!")
 
+method connect*(
+  self: Dial,
+  addrs: seq[MultiAddress]): Future[PeerId] {.async, base.} =
+  ## Connects to a peer and retrieve it's PeerId
```
Comment (arnetheduck) [inline libp2p/dial.nim:37]:
```suggestion
  ## Connects to a peer and retrieve its PeerId
```

File: libp2p/dialer.nim [hunk 2]
```diff
@@ -101,20 +98,22 @@ proc dialAndUpgrade(
...
   addrs: seq[MultiAddress],
   forceDial: bool):
   Future[Connection] {.async.} =
-  if self.localPeerId == peerId:
+  if Opt.some(self.localPeerId) == peerId:
     raise newException(CatchableError, "can't dial self!")
 
   # Ensure there's only one in-flight attempt per peer
-  let lock = self.dialLock.mgetOrPut(peerId, newAsyncLock())
+  let lock = self.dialLock.mgetOrPut(peerId.get(default(PeerId)), newAsyncLock())
```
Comment (arnetheduck) [inline libp2p/dialer.nim:109]: would it make sense to block on addresses here instead?

File: libp2p/dialer.nim [hunk 3]
```diff
@@ -101,20 +98,22 @@ proc dialAndUpgrade(
...
   try:
     await lock.acquire()
 
     # Check if we have a connection already and try to reuse it
-    var conn = self.connManager.selectConn(peerId)
+    var conn =
```
Comment (arnetheduck) [inline libp2p/dialer.nim:114]: where do we check that after a discovery of peerid, we don't have too many connections to the same peer?

Comment (Menduist) [reply libp2p/dialer.nim:114]: https://github.com/status-im/nim-libp2p/blob/e6440c43c211d930d72d6189d4bca67fef7276b3/libp2p/connmanager.nim#L387-L390

**Review by arnetheduck — COMMENTED** [review]
lgtm, can't approve my own PR :)

File: libp2p/dialer.nim (hunk 2 above)
Comment (Menduist) [reply libp2p/dialer.nim:109]: Don't have a strong opinion, though it seems more tricky to do

**Pull request merged**